<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Tade Souaiaia" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Day 4b - prsWorkshop</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Day 4b";
        var mkdocs_page_input_path = "Day4b.docx.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]--> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> prsWorkshop
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Software</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../quik_install/">Checklist</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quik_install/">Checklist</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quik_demo/">Demos</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Program</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../modules/Day1a.docx/">Day 1a</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day1b.docx/">Day 1b</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day2.docx/">Day 2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day3a.docx/">Day 3a</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day3b.docx/">Day 3b</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day4a.docx/">Day 4a</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Day 4b</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#module-goals">Module Goals</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#part1-plot-decay-of-ancestry-ld-over-time">Part1: Plot Decay of Ancestry LD over time</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#questions">Questions</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#i-describe-what-happens-to-admixture-ld-over-time">(i) Describe what happens to admixture LD over time?</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#ii-why-does-recombination-also-have-an-impact">(ii) Why does recombination also have an impact?</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#part-2-global-ancestry-inference">Part 2: Global Ancestry Inference</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day5_Projects/">Day5</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../about_future/">Future Work</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">prsWorkshop</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Program</li>
      <li class="breadcrumb-item active">Day 4b</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/tadesouaiaia/prsWorkshop-website/edit/master/docs/Day4b.docx.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="day-4-practical-2-introduction-to-admixture-analysis">Day 4 - Practical 2: Introduction to Admixture analysis</h2>
<h3 id="module-goals">Module Goals</h3>
<p>The goal of this practical is to provide you with basic understanding of Admixture and the basic elements behind Admixture PRS scores. Upon completion of this practical, you should:
* Gain familiarity with a variety of tools used in the context of admixed population research
* Go through the basic steps of formulating admixture-informed polygenic risk scores</p>
<h3 id="part1-plot-decay-of-ancestry-ld-over-time">Part1: Plot Decay of Ancestry LD over time</h3>
<p>Here we explore how Admixture LD varies over time and as a function of the genetic distance between loci.
In R:</p>
<pre><code>library(ggplot2)
library(reshape2)
library(viridis)

#range of values of r (recombination fraction)
r = seq(0, 0.5, 0.1)
dtmat = matrix(NA, nrow = 6, ncol = 10) # matrix to store values of dt


for(i in 1:6){
  dt = 0.25
  for(j in 1:10){
    dtmat[i,j] = dt
    dt = dt*(1 -r[i])^j
  }
}

dtmat = reshape2::melt(dtmat)
colnames(dtmat) = c(&quot;r&quot;,&quot;g&quot;,&quot;Dt&quot;)
dtmat$r = (dtmat$r - 1)/10

ggplot(dtmat, 
       aes(x = g, y = Dt, group = r, color = as.factor(r))) +
  geom_line(linewidth = 1.2) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text = element_text(size = 14),
    axis.title = element_text(size = 16),
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12),
    plot.title = element_text(size = 18, face = &quot;bold&quot;),
    plot.subtitle = element_text(size = 14),
    panel.grid.major = element_line(color = &quot;gray80&quot;)
  ) +
  scale_color_viridis_d() +
  scale_x_continuous(breaks = c(1:10)) +
  labs(
    title = &quot;Decay of Linkage Disequilibrium Over Generations&quot;,
    subtitle = &quot;Effect of Recombination Fraction (r) on Admixture LD&quot;,
    x = &quot;Generations since admixture (t)&quot;,
    y = &quot;Admixture LD&quot;,
    color = &quot;Recombination Fraction (r)&quot;
  )
</code></pre>
<h4 id="questions"><strong>Questions</strong></h4>
<h5 id="i-describe-what-happens-to-admixture-ld-over-time">(i) Describe what happens to admixture LD over time?</h5>
<h5 id="ii-why-does-recombination-also-have-an-impact">(ii) Why does recombination also have an impact?</h5>
<h3 id="part-2-global-ancestry-inference">Part 2: Global Ancestry Inference</h3>
<p>We will now run an analysis using the software ADMIXTURE to calculate global ancestry proportions across a sample of 28 individuals. Here we perform a supervised analysis. Please execute the following code from location ~/RFMIX_WCS_2024/data/plink/
 ```
 ./admixture samples_n28_qc_thin.bed 2 --supervised -j4 </p>
<pre><code>#### **Questions**
##### (i) What do you think the number specified after the inout file represents?

The next step is to create a plot the results
</code></pre>
<h1 id="plot-global-ancestry-results">Plot Global Ancestry Results</h1>
<p>R
 library(ggplot2)
 library(reshape2)</p>
<p># Read the data from files
 fam &lt;- read.table("samples_n28_qc_thin.fam", header = FALSE)
 pop &lt;- read.table("samples_n28_qc_thin.pop", header = FALSE)
 Q &lt;- read.table("samples_n28_qc_thin.2.Q", header = FALSE)</p>
<p># Merge the data into one data frame
 merged_data &lt;- cbind(fam, pop, Q)</p>
<p># Extract necessary columns (assuming the structure of the files as in the image)
 colnames(merged_data) &lt;- c("FID", "IID", "PaternalID", "MaternalID", "Sex", "Phenotype", "Population", "AFR", "EUR")</p>
<p># Order data by decreasing AFR
 merged_data &lt;- merged_data[order(-merged_data$AFR), ]</p>
<p># Prepare data for plotting
 plot_data &lt;- merged_data[, c("IID", "AFR", "EUR")]
 plot_data$IID &lt;- factor(plot_data$IID, levels = plot_data$IID)  # Ensure order is maintained in plot
 plot_data_melted &lt;- melt(plot_data, id.vars = "IID")</p>
<p># Create the plot
 p &lt;- ggplot(plot_data_melted, aes(x = IID, y = value, fill = variable)) +
   geom_bar(stat = "identity", position = "stack") +
   scale_fill_manual(values = c("AFR" = "red", "EUR" = "darkgreen")) +
   labs(x = "Individual ID", y = "Ancestry Proportion", fill = "Ancestry", title = "Global Ancestry Proportions") +
   theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 10, face = "bold"),
         axis.text.y = element_text(face = "bold"),  # Bold y-axis numbers
         axis.title.x = element_text(face = "bold"),
         axis.title.y = element_text(face = "bold"),
         legend.title = element_text(face = "bold"),
         plot.title = element_text(face = "bold", hjust = 0.5))</p>
<p># Save the plot with specified dimensions
 ggsave("ancestry_plot.png", plot = p, width = 10, height = 6, units = "in", dpi = 300)</p>
<pre><code>#### Questions
##### (i) What are the ancestry assignments of the 28 individuals? (Provide estimated proportions where necessary)


### Part 3: Local Ancestry Inference 

Next we will use the RFMix software to calculate local ancestry on chromosome 22 for the same 28 individuals. The RFMix algorithm uses an unsupervised learning algorithm.
</code></pre>
<h1 id="run-the-following-unix-command-from-the-home-directory">Run the following UNIX command from the home directory</h1>
<p>module load cmake
module load bcftools/1.10.2</p>
<p>for i in {22..22}; do<br />
./software/rfmix -f ./data/rfmix/chr1-22_phased.bcf.gz -r ./reference/chr1-22.b.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.bcf.gz --analyze-range=26.86-31.80 -m ./data/rfmix/1KG_superpop_vs_ID.txt --chromosome=${i} -g ./reference/1kg_chr1-22.gmap --n-threads=4 -o ./out/rfmix/chr${i}.local_ancestry; done</p>
<pre><code>
#### Questions
##### (i) What information is provided in the simulation results table displayed on-screen ?


### Part 4: Plot local admixture on chromosome 22

In this step we will plot the output from the previous RFMix run. Execute the following code from the home directory

</code></pre>
<p>R
library(ggplot2)
library(dplyr)
library(tidyr)</p>
<h1 id="function-to-read-the-msptsv-file">Function to read the .msp.tsv file</h1>
<p>read_msp_file &lt;- function(msp_file) {
  # Read the entire file as text
  lines &lt;- readLines(msp_file)</p>
<p># Determine the number of header lines (lines starting with '#')
  header_lines &lt;- lines[grepl("^#", lines)]
  num_header_lines &lt;- length(header_lines)</p>
<p># Extract column names from the last header line
  col_names &lt;- strsplit(header_lines[num_header_lines], "\t")[[1]]</p>
<p># Read the data skipping the header lines
  msp_df &lt;- read.table(msp_file, header = FALSE, skip = num_header_lines, sep = "\t")</p>
<p># Assign column names
  colnames(msp_df) &lt;- col_names
  return(msp_df)
}</p>
<h1 id="function-to-read-the-q-file">Function to read the .Q file</h1>
<p>read_Q_file &lt;- function(Q_file) {
  Q_df &lt;- read.table(Q_file, header = TRUE, sep = "\t", comment.char = "#")
  colnames(Q_df) &lt;- c("sample", "AFR", "AMR", "EAS", "EUR", "SAS")
  return(Q_df)
}</p>
<h1 id="specification-of-file-paths">Specification of file paths</h1>
<p>msp_file &lt;- './out/rfmix/chr22.local_ancestry.msp.tsv'
Q_file &lt;- './out/rfmix/chr22.local_ancestry.rfmix.Q'</p>
<h1 id="read-in-files">Read in files</h1>
<p>msp_df &lt;- read_msp_file(msp_file)
Q_df &lt;- read_Q_file(Q_file)</p>
<h1 id="check-for-na-or-empty-column-names">Check for NA or empty column names</h1>
<p>print(colnames(msp_df))</p>
<h1 id="extract-ancestry-columns-based-on-pattern-recognition-of-column-names">Extract ancestry columns based on pattern recognition of column names</h1>
<p>ancestry_cols &lt;- colnames(msp_df)[grep("\.0$|\.1$", colnames(msp_df))]</p>
<h1 id="function-to-determine-ancestry-based-on-rfmix-codes">Function to determine ancestry based on RFMix codes</h1>
<p>determine_ancestry &lt;- function(value) {
  ancestry_map &lt;- c("AFR", "AMR", "EAS", "EUR", "SAS")
  return(ancestry_map[value + 1])
}</p>
<h1 id="rename-columns-uniquely">Rename columns uniquely</h1>
<p>colnames(msp_df)[7:62] &lt;- paste0("haplo_", colnames(msp_df)[7:62])</p>
<h1 id="prepare-data-for-plotting">Prepare data for plotting</h1>
<p>plot_data &lt;- msp_df %&gt;%
  pivot_longer(cols = starts_with("haplo_"), names_to = "haplo_col", values_to = "ancestry_code", names_repair = "unique") %&gt;%
  mutate(
    individual = gsub("\..*", "", haplo_col),
    strand = ifelse(grepl("\.0$", haplo_col), "Strand 1", "Strand 2"),
    ancestry = determine_ancestry(ancestry_code)
  )</p>
<h1 id="create-plot-of-chromosome-22-across-the-sample">Create plot of chromosome 22 across the sample</h1>
<p>plot &lt;- ggplot(plot_data, aes(x = sgpos, xend = egpos, y = interaction(individual, strand, lex.order = TRUE), yend = interaction(individual, strand, lex.order = TRUE), color = ancestry)) +
  geom_segment(linewidth = 4) +
  scale_color_manual(values = c("AFR" = "blue", "AMR" = "orange", "EAS" = "green", "EUR" = "red", "SAS" = "purple")) +
  labs(x = "Genetic position (cM)", y = "Individual ID and Haplotype", title = "Local Ancestry Across Chromosome 22", color = "Ancestry") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 8), axis.title.y = element_text(size = 10), axis.title.x = element_text(size = 10))</p>
<h1 id="save-the-plot-to-a-file">Save the plot to a file</h1>
<p>ggsave("./out/rfmix/local_ancestry_chromosome22_5Mb_subregion.png", plot = plot, width = 10, height = 8)</p>
<pre><code>#### Questions
##### (i) How many different continental ancestries do you see represented across the 58 strands?
##### (ii) Why is the number of different ancestry backgrounds higher than it was in the previous step?


### Part 5: Formatting of admixture files for analysis using PRSice

#### Step 1 - Convert phased genotypes to GenomicRange format
</code></pre>
<h1 id="run-from-the-home-directory">Run from the home directory</h1>
<p>library(vcfR)
library(memuse)
library(panelr)
library(GenomicRanges)
library(data.table)</p>
<p>clean_memory &lt;- function(vars_to_remove) {
  rm(list = vars_to_remove)
  gc()
}</p>
<h1 id="read-in-phased-vcf-file">Read in phased VCF file</h1>
<p>vcfr_inputfile_chr22 &lt;- read.vcfR("./data/rfmix/chr22_phased.vcf.gz", verbose = FALSE)
extracted_haps22 &lt;- extract.haps(vcfr_inputfile_chr22, mask = FALSE, unphased_as_NA = TRUE, verbose = TRUE)
extracted_snp_info22 &lt;- getFIX(vcfr_inputfile_chr22)</p>
<h1 id="convert-haplotypes-to-data-frame">Convert haplotypes to data frame</h1>
<p>haps_df &lt;- data.frame(extracted_haps22, check.names = FALSE)
haps_dt &lt;- setDT(haps_df, keep.rownames = "snps")</p>
<h1 id="convert-snp-info-to-data-frame">Convert SNP info to data frame</h1>
<p>snp_info_df &lt;- data.frame(extracted_snp_info22)</p>
<h1 id="check-for-numerical-and-ordering-consistency-between-haplotypes-and-snp-info">Check for numerical and ordering consistency between haplotypes and SNP info</h1>
<p>if (!identical(haps_dt[['snps']], snp_info_df[['ID']])) {
  stop("Numerical and ordering inconsistency between haplotypes and SNP info")
}</p>
<h1 id="merge-haplotypes-and-snp-info">Merge haplotypes and SNP info</h1>
<p>merge_chr22 &lt;- cbind(snp_info_df, haps_dt)</p>
<h1 id="convert-to-long-format-using-panelr">Convert to long format using panelr</h1>
<p>chr22_haplo_long &lt;- long_panel(merge_chr22, prefix = "_", begin = 0, end = 1, label_location = "end", as_panel_data = FALSE)</p>
<h1 id="insert-end-column-after-pos-and-create-wave-column">Insert 'end' column after 'POS' and create 'wave' column</h1>
<p>chr22_haplo_long$end &lt;- chr22_haplo_long$POS
chr22_haplo_long$wave &lt;- ifelse(chr22_haplo_long$wave == "0", "+", ifelse(chr22_haplo_long$wave == "1", "-", "Z"))</p>
<h1 id="remove-row-names-and-drop-redundant-columns-using-base-r">Remove row names and drop redundant columns using base R</h1>
<p>rownames(chr22_haplo_long) &lt;- NULL
drops &lt;- c("id", "snps", "QUAL", "FILTER")
chr22_haplo_long &lt;- chr22_haplo_long[, !names(chr22_haplo_long) %in% drops]</p>
<h1 id="rename-columns">Rename columns</h1>
<p>names(chr22_haplo_long)[3] &lt;- "start"
names(chr22_haplo_long)[1] &lt;- "strand"
header &lt;- gsub(".*_", "", colnames(chr22_haplo_long)[7:ncol(chr22_haplo_long)])
names(chr22_haplo_long)[7:ncol(chr22_haplo_long)] &lt;- header</p>
<h1 id="create-granges-object">Create GRanges object</h1>
<p>gr_obj_chr22 &lt;- makeGRangesFromDataFrame(chr22_haplo_long, ignore.strand = FALSE, keep.extra.columns = TRUE)</p>
<h1 id="save-the-granges-object">Save the GRanges object</h1>
<p>saveRDS(gr_obj_chr22, file = "./out/rfmix/chr22_phased_gr.rds")</p>
<h1 id="clean-up-memory">Clean up memory</h1>
<p>clean_memory(c("vcfr_inputfile_chr22", "extracted_haps22", "merge_chr22", "haps_df", "haps_dt", "snp_info_df", "chr22_haplo_long", "gr_obj_chr22"))</p>
<pre><code>
### Part 5: Step 2 - Merge genotypes from Step 1 with local ancestry calls by RFMix
</code></pre>
<h1 id="read-in-msp-file">Read in MSP file</h1>
<p>msp &lt;- fread("./out/rfmix/chr22.local_ancestry.msp.tsv")</p>
<h1 id="reformat-genome-wide-local-ancestry-output-as-granges-object">Reformat genome-wide local ancestry output as GRanges object</h1>
<p>colnames(msp)[1] &lt;- "chm"
msp_gr &lt;- makeGRangesFromDataFrame(msp,
                                   seqnames.field = "chm",
                                   start.field = "spos",
                                   end.field = "epos",
                                   keep.extra.columns = TRUE)</p>
<h1 id="convert-the-elements-of-the-grange-object-into-a-dataframe">Convert the elements of the GRange object into a dataframe</h1>
<p>chr22_rf &lt;- data.frame(seqnames = seqnames(msp_gr),
                       ranges = ranges(msp_gr),
                       strand = strand(msp_gr),
                       mcols(msp_gr), check.names = FALSE)
names(chr22_rf)[1:5] &lt;- c("chr", "start", "end", "width", "strand")
rownames(chr22_rf) &lt;- NULL</p>
<h1 id="clean-up-column-headers">Clean up column headers</h1>
<p>header &lt;- gsub(".*_", "", colnames(chr22_rf)[9:ncol(chr22_rf)])
names(chr22_rf)[9:ncol(chr22_rf)] &lt;- header</p>
<h1 id="convert-local-ancestry-calls-from-wide-to-long-format">Convert local ancestry calls from wide to long format</h1>
<p>chr22_rf_long &lt;- long_panel(chr22_rf, prefix = ".", begin = 0, end = 1, label_location = "end", as_panel_data = FALSE)
chr22_rf_long &lt;- as.data.frame(chr22_rf_long, check.names = FALSE)
chr22_rf_long$strand &lt;- ifelse(chr22_rf_long$wave == "0", "+", ifelse(chr22_rf_long$wave == "1", "-", "Z"))</p>
<h1 id="drop-redundant-columns">Drop redundant columns</h1>
<p>drops &lt;- c("wave", "id")
chr22_rf_long &lt;- chr22_rf_long[, !(names(chr22_rf_long) %in% drops)]
rownames(chr22_rf_long) &lt;- NULL</p>
<h1 id="convert-the-reconfigured-dataframe-file-into-a-granges-object">Convert the reconfigured dataframe file into a GRanges object</h1>
<p>chr22_msp_gr &lt;- makeGRangesFromDataFrame(chr22_rf_long, ignore.strand = FALSE, keep.extra.columns = TRUE)</p>
<h1 id="ensure-chr22_haplo_long-is-a-granges-object-before-finding-overlaps">Ensure chr22_haplo_long is a GRanges object before finding overlaps</h1>
<p>chr22_haplo_gr &lt;- makeGRangesFromDataFrame(chr22_haplo_long, ignore.strand = FALSE, keep.extra.columns = TRUE)</p>
<h1 id="find-overlaps-and-store-matching-features">Find overlaps and store matching features</h1>
<h1 id="matching-is-coordinates-based-base-position-startendis-used-to-match-the-two-files">Matching is coordinates based. Base position ("start"/"end")is used to match the two files</h1>
<p>matched_regions &lt;- findOverlaps(chr22_haplo_gr, chr22_msp_gr)
chr22_haps_lanc_gr &lt;- chr22_haplo_gr[queryHits(matched_regions)]  #Store matching features in a new dataframe, add metadata from RFmix output.
mcols(chr22_haps_lanc_gr) &lt;- cbind.data.frame(mcols(chr22_haps_lanc_gr), mcols(chr22_msp_gr[subjectHits(matched_regions)]))</p>
<h1 id="local-ancestry-calls-are-now-aligned-with-genotypic-data-and-positional-coordinates">Local ancestry calls are now aligned with genotypic data and positional coordinates</h1>
<h1 id="convert-to-dataframe-without-adding-x-to-numeric-column-names">Convert to dataframe without adding 'X' to numeric column names</h1>
<p>genes_df &lt;- as.data.frame(chr22_haps_lanc_gr)
names(genes_df) &lt;- sub('^X', '', names(genes_df))</p>
<h1 id="output-the-dataframe">Output the dataframe</h1>
<p>write.table(genes_df, file = "./out/rfmix/chr22_phased_geno_lanc.txt", quote = FALSE, row.names=F)</p>
<h1 id="clean-up-memory_1">Clean up memory</h1>
<p>rm(list = c("chr22_haplo_gr", "chr22_haplo_long", "msp", "msp_gr", "chr22_rf", "header",
            "chr22_rf_long", "chr22_msp_gr", "matched_regions",
            "chr22_haps_lanc_gr"))
gc()</p>
<pre><code>
#### Part 5: Step 3 - Create separate Plink files
</code></pre>
<h1 id="partition-genotype-and-local-ancestry-data">Partition genotype and local ancestry data</h1>
<p>chr22_genos &lt;- genes_df[, c(1, 2, 6, 9:36)]
chr22_LA &lt;- genes_df[, c(1, 2, 6, 40:ncol(genes_df))]</p>
<h1 id="reintegrate-in-interleaved-format">Reintegrate in interleaved format</h1>
<p>d &lt;- chr22_genos[, -c(1:3)]  # Retain ID columns only
d2 &lt;- chr22_LA[, -c(1:3)]</p>
<h1 id="prepare-headers">Prepare headers</h1>
<p>indx &lt;- rbind(names(d), names(d2))
dmerge &lt;- cbind(d, d2)
dfinal &lt;- dmerge[, indx]</p>
<h1 id="convert-lanc-data-from-long-to-wide-format">Convert LAnc data from long to wide format</h1>
<p>LA_wide &lt;- lapply(1:ncol(chr22_LA), function(i) as.data.frame(matrix(chr22_LA[, i], ncol = 2, byrow = TRUE)))</p>
<h1 id="check-length-of-la_wide">Check length of LA_wide</h1>
<p>length(LA_wide)  # Expected number. Each item contains a set of matched strand pairs per individual</p>
<p>LA_final &lt;- as.data.frame(do.call(cbind, LA_wide))
LA_final &lt;- LA_final[, -c(2, 4, 6)]  # Remove duplicates of first 3 cols
colnames(LA_final)[4:ncol(LA_final)] &lt;- colnames(dfinal)  # Apply headers</p>
<h1 id="convert-genotype-calls-to-horizontal-orientation">Convert genotype calls to horizontal orientation</h1>
<p>geno_wide &lt;- lapply(1:ncol(chr22_genos), function(i) as.data.frame(matrix(chr22_genos[, i], ncol = 2, byrow = TRUE)))  # Genotypic part: Get horizontal genotypes
geno_final &lt;- as.data.frame(do.call(cbind, geno_wide))  # Merge horizontal genotypes</p>
<h1 id="clean-up-and-finalize-geno_final">Clean up and finalize geno_final</h1>
<p>geno_final &lt;- geno_final[, -c(2, 4, 6)]  # Redundant columns
colnames(geno_final)[4:ncol(geno_final)] &lt;- colnames(dfinal)</p>
<h1 id="name-columns">Name columns</h1>
<p>colnames(geno_final)[1:3] &lt;- c("CHROM", "BP", "ID")
colnames(LA_final)[1:3] &lt;- c("CHROM", "BP", "ID")</p>
<h1 id="write-final-tables">Write final tables</h1>
<p>write.table(geno_final, "./out/rfmix/chr22_geno.txt", row.names = FALSE, quote = FALSE)
write.table(LA_final, "./out/rfmix/chr22_LA.txt", row.names = FALSE, quote = FALSE)</p>
<pre><code>
#### Part 5: Step 4 - Use custom software (RFTransform) to create Plink files for input into PRSice
</code></pre>
<p>module load cmake
./software/RFTransform/build/RFTransformer ./out/rfmix/chr22_geno.txt ./out/rfmix/chr22_LA.txt ./out/plink/chr22</p>
<h1 id="perform-general-qc-ahead-of-running-prsice-on-ancestry-deconvolved-individuals">Perform general QC ahead of running PRSice on ancestry-deconvolved individuals</h1>
<h1 id="i-remove-snps-with-low-minor-allele-count-mac">(i) Remove SNPs with low minor allele count (MAC)</h1>
<h1 id="plink-remove-monomorphic-snps-minor-allele-count-0-4">Plink - Remove monomorphic SNPs (minor allele count 0-4).</h1>
<h1 id="afr">AFR</h1>
<p>for i in {22..22}; do
        ./software/plink2 \
        --bfile ./out/plink/chr${i}-AFR \
        --mac 5 \
        --make-bed \
        --out ./out/plink/chr${i}-AFR.mac
done</p>
<h1 id="eur">EUR</h1>
<p>for i in {22..22}; do
        ./software/plink2 \
        --bfile ./out/plink/chr${i}-EUR \
        --mac 5 \
        --make-bed \
        --out ./out/plink//chr${i}-EUR.mac
done</p>
<h1 id="prsice-generate-ancestry-specific-weights">PRSice - Generate ancestry-specific weights</h1>
<h1 id="afr_1">AFR</h1>
<p>Rscript ./software/PRSice.R \
--prsice ./software/PRSice_linux \
--base ./data/plink/AFR-BMI.Phenotype.glm.linear \
--extract ./data/plink/snp.valid \
--A1 A1 \
--pvalue P \
--stat BETA \
--pheno ./data/plink/pheno.plink \
--beta \
--snp ID \
--score sum \
--target ./out/plink/chr22-AFR.mac \
--binary-target F \
--out ./out/prsice/BMI_AFR-base</p>
<h1 id="eur_1">EUR</h1>
<p>Rscript ./software/PRSice.R \
--prsice ./software/PRSice_linux \
--base ./data/plink/EUR-BMI.Phenotype.glm.linear \
--extract ./data/plink/snp.valid \
--A1 A1 \
--pvalue P \
--stat BETA \
--pheno ./data/plink/pheno.plink \
--beta \
--snp ID \
--score sum \
--target ./out/plink/chr22-EUR.mac \
--binary-target F \
--out ./out/prsice/BMI_EUR-base</p>
<pre><code>
#### Part 5: Step 5 - Evaluate the Admixture-informed PRS
</code></pre>
<p>library(dplyr)</p>
<h1 id="read-the-prs-files-into-dataframes">Read the PRS files into dataframes</h1>
<p>file1 &lt;- read.table("./out/prsice/BMI_EUR-base.best", header = TRUE, check.names=F)
file2 &lt;- read.table("./out/prsice/BMI_AFR-base.best", header = TRUE, check.names=F)</p>
<h1 id="add-the-fourth-column-of-both-files">Add the fourth column of both files</h1>
<p>file1$PRS_SUM &lt;- file1$PRS + file2$PRS</p>
<h1 id="load-the-phenotype-data">Load the phenotype data</h1>
<p>pheno &lt;- read.table("./data/plink/pheno.plink", header = F)
colnames(pheno) &lt;- c("FID", "IID", "phenotype")</p>
<h1 id="convert-phenotype-column-to-numeric">Convert phenotype column to numeric</h1>
<p>pheno$phenotype &lt;- as.numeric(pheno$phenotype)</p>
<h1 id="merge-prs-data-with-phenotype-data">Merge PRS data with phenotype data</h1>
<p>merged_data &lt;- merge(file1[, c("FID", "IID", "PRS", "PRS_SUM")], pheno, by = c("FID", "IID"))
merged_data &lt;- merge(merged_data, file2[, c("FID", "IID", "PRS")], by = c("FID", "IID"))</p>
<h1 id="rename-columns_1">Rename columns</h1>
<p>names(merged_data)[names(merged_data) == "PRS.x"] &lt;- "PRS1"
names(merged_data)[names(merged_data) == "PRS.y"] &lt;- "PRS2"</p>
<h1 id="perform-linear-regression-for-each-prs-and-the-combined-prs">Perform linear regression for each PRS and the combined PRS</h1>
<p>model1 &lt;- lm(phenotype ~ PRS1, data = merged_data)
model2 &lt;- lm(phenotype ~ PRS2, data = merged_data)
model_sum &lt;- lm(phenotype ~ PRS_SUM, data = merged_data)</p>
<h1 id="extract-r-squared-values">Extract R-squared values</h1>
<p>r_squared1 &lt;- summary(model1)$r.squared
r_squared2 &lt;- summary(model2)$r.squared
r_squared_sum &lt;- summary(model_sum)$r.squared</p>
<h1 id="print-the-r-squared-values">Print the R-squared values</h1>
<p>cat("R-squared for PRS1: ", r_squared1, "\n")
cat("R-squared for PRS2: ", r_squared2, "\n")
cat("R-squared for PRS_SUM: ", r_squared_sum, "\n")
```</p>
<h4 id="questions_1"><strong>Questions</strong></h4>
<h5 id="i-how-does-the-r-squared-of-the-combined-ancestry-prs-perform-relative-to-the-2-partial-genome-prss">(i) How does the R-squared of the combined-ancestry PRS perform relative to the 2 partial-genome PRSs?</h5>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../Day4a.docx/" class="btn btn-neutral float-left" title="Day 4a"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../Day5_Projects/" class="btn btn-neutral float-right" title="Day5">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/tadesouaiaia/prsWorkshop-website" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../Day4a.docx/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../Day5_Projects/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
