<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Tade Souaiaia" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Advanced Polygenic Risk Score Analyses - prsWorkshop</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Advanced Polygenic Risk Score Analyses";
        var mkdocs_page_input_path = "Day2b.docx.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> prsWorkshop
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Preparation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../prep_list/">Checklist</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../prep_os/">System</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../prep_terminal/">Terminal</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../prep_software/">Software</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../prep_testing/">Testing</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Program</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../Day2.docx/">Day 1a</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day3a.docx/">Day 1b</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day3b.docx/">Day 2a</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Day4a.docx/">Day 2b</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorials</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../misc_commands/">Bash Tutorial</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../misc_R/">R Tutorial</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../misc_python/">Python Tutorial</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../misc_plink/">Plink Tutorial</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Misc</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../misc_linux/">Installing Linux</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../misc_plink_problem/">Mac Security</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">prsWorkshop</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Advanced Polygenic Risk Score Analyses</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/tadesouaiaia/prsWorkshop-website/edit/master/docs/Day2b.docx.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="advanced-polygenic-risk-score-analyses">Advanced Polygenic Risk Score Analyses</h1>
<h2 id="day-2-afternoon-practical-pathway-prs-analyses">Day 2 - Afternoon Practical: Pathway PRS Analyses</h2>
<h2 id="table-of-contents">Table of Contents</h2>
<ol>
<li><a href="#gene-set-analysis-intro">Introduction to gene set (pathway) PRS analysis</a></li>
<li><a href="#prset-inputs">Inputs required for gene-set PRS analysis</a><ol>
<li><a href="#molecular-signatures-Database-msigdb">Molecular Signatures Database MSigDB</a></li>
<li><a href="#other-inputs">Other inputs that can be used for gene-set PRS using PRSet</a></li>
</ol>
</li>
<li><a href="#considerations">Considerations when analysing and interpreting gene-set PRSs</a><ol>
<li><a href="#clumping">Clumping in gene set PRS analyses</a></li>
<li><a href="#thresholding">P-value thresholding in gene set PRS analyses</a></li>
<li><a href="#p-value-testing">Self-Contained vs Competitive Testing</a></li>
</ol>
</li>
<li><a href="#exercise-4-gene-set-based-prs-analysis">Exercise: Calculate gene-set PRS analysis</a></li>
</ol>
<h2 id="key-learning-outcomes">Key Learning Outcomes</h2>
<p>After completing this practical, you should be able to:
  1. Understand the motivation and rationale for calculating gene-set PRS.
  2. Identify the additional inputs required for gene-set PRS analysis.
  3. Differentiate between self-contained and competitive tests in gene-set PRS analyses.
  4. Interpret the outcomes of gene-set PRSs and how they differ from genome-wide PRS.
  5. Calculate gene-set based PRSs using PRSet.</p>
<h2 id="resources-you-will-be-using">Resources you will be using</h2>
<p>Similar to standard genome-wide PRS analyses, summary statistics from Genome-Wide Association Studies (GWAS) and individual level genotype and phenotype data are required to perform gene-set PRS analyses. In this session, the following datasets are used:</p>
<table>
<thead>
<tr>
<th style="text-align: center;"><strong>Phenotype</strong></th>
<th style="text-align: center;"><strong>Provider</strong></th>
<th style="text-align: center;"><strong>Description</strong></th>
<th style="text-align: center;"><strong>Download Link</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">Height</td>
<td style="text-align: center;"><a href="https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files">GIANT Consortium</a></td>
<td style="text-align: center;">GWAS of height on 253,288 individuals</td>
<td style="text-align: center;"><a href="https://portals.broadinstitute.org/collaboration/giant/images/0/01/GIANT_HEIGHT_Wood_et_al_2014_publicrelease_HapMapCeuFreq.txt.gz">Link</a></td>
</tr>
</tbody>
</table>
<p>Additionally, to perform gene-set level analyses, information about the genomic regions for which we want to calculate the PRSs are required. This information can be obtained from the following database:</p>
<table>
<thead>
<tr>
<th style="text-align: center;"><strong>Data Set</strong></th>
<th style="text-align: center;"><strong>Description</strong></th>
<th style="text-align: center;"><strong>Download Link</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">Ensembl Human Genome GTF file</td>
<td style="text-align: center;">A file containing the coordinates for genes in the human genome. Used by PRSet to map the SNPs onto genic regions</td>
<td style="text-align: center;"><a href="https://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/">Link</a></td>
</tr>
<tr>
<td style="text-align: center;">MSigDB Gene Sets</td>
<td style="text-align: center;">File containing the gene-set information. <em>Free registration required.</em></td>
<td style="text-align: center;"><a href="http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/6.1/h.all.v6.1.symbols.gmt">Download here after registration</a></td>
</tr>
</tbody>
</table>
<p><a href="#top"><a href="#table-of-contents">Back to Top</a></a></p>
<p><a id="gene-set-analysis-intro"></a></p>
<h2 id="introduction-to-gene-set-pathway-prs-analysis">Introduction to gene set (pathway) PRS analysis</h2>
<p>Currently, most PRS analyses have been performed on a genome-wide scale. This does not account for the substructure of the genome.</p>
<p>Here we will learn how to run a gene set (or pathway) based PRS analyses. The key difference between genome-wide PRS and gene set or pathway-based PRSs analyses is that, instead of aggregating the estimated effects of risk alleles across the entire genome, gene-set PRSs aggregate risk alleles across k gene sets separately.</p>
<p><img alt="pathway PRS" src="https://github.com/tadesouaiaia/prsWorkshop-website/blob/main/docs/images/pathwayPRS_overview.png" /></p>
<p>Gene-set PRS analyses may account for genomic sub-structure, constitute an extension to the classic polygenic model of disease, and may better reflect disease heterogeneity. For more information about the rationale and the software that we are going to use, please see the PRSet publication <a href="https://doi.org/10.1371/journal.pgen.1010624">PRSet: Pathway-based polygenic risk score analyses and software</a>. </p>
<p>In this practical, we will go through some of the additional input requirements and considerations for the analysis of gene-set PRS analysis and will then calculate some gene-set based PRS using <a href="https://choishingwan.github.io/PRSice/quick_start_prset/">PRSet</a>.</p>
<p><a id="prset-inputs"></a></p>
<h2 id="inputs-required-for-gene-set-prs-analysis">Inputs required for gene-set PRS analysis</h2>
<p>PRSet is based on PRSice, but with some addtional input requirements about the gene-sets for which PRSs are calculated. In this tutorial, we will use as input gene-sets from the <strong>Molecular Signatures Database</strong>. However, PRSet also takes as input <strong>BED and SNP files</strong>. </p>
<p><a id="molecular-signatures-Database-msigdb"></a></p>
<h3 id="molecular-signatures-database-msigdb">Molecular Signatures Database MSigDB</h3>
<p>MSigDB oÔ¨Äers an excellent source of gene-sets, including the hallmark genes, gene-sets of diÔ¨Äerent biological processes, gene-sets of diÔ¨Äerent oncogenic signatures etc. All gene-sets from MSigDB follows the Gene Matrix Transposed file format (GMT), which consists of one line per gene-set, each containing at least 3 column of data:</p>
<table>
<thead>
<tr>
<th style="text-align: center;"></th>
<th style="text-align: center;"></th>
<th style="text-align: center;"></th>
<th style="text-align: center;"></th>
<th style="text-align: center;"></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">Set A</td>
<td style="text-align: center;">Description</td>
<td style="text-align: center;">Gene 1</td>
<td style="text-align: center;">Gene 2</td>
<td style="text-align: center;">...</td>
</tr>
<tr>
<td style="text-align: center;">Set A</td>
<td style="text-align: center;">Description</td>
<td style="text-align: center;">Gene 1</td>
<td style="text-align: center;">Gene 2</td>
<td style="text-align: center;">...</td>
</tr>
</tbody>
</table>
<hr />
<blockquote>
<p>** Have a look at the Reference/Sets.gmt file. **</p>
<p>‚ùì How many gene-sets are there in the Reference/Sets.gmt file? </p>
<p>‚ùì How many genes does the largest gene-set contain?
</p>
</blockquote>
<hr />
<blockquote>
<p>üí¨ While you can read the GMT file using Excel. You should be aware that Excel has a tendency to convert gene names into dates (e.g. SEPT9 to Sep-9)
</p>
</blockquote>
<hr />
<p>As GMT format does not contain the chromosomal location for each individual gene, an additional file (General Transfer Format file) is required to provide the chromosoaml location such that SNPs can be map to genes.</p>
<h3 id="general-transfer-format-file">General Transfer Format file</h3>
<p>The General Transfer Format (GTF) file contains the chromosomal coordinates for each gene. It is a <strong>tab</strong> separated file and all but the final field in each feature line must contain a value. "Empty" columns should be denoted with a ‚Äò.‚Äô. You can read the full format specification here. One column that might be of particular interest is column 3: <strong>feature</strong>, which indicates what feature that line of GTF represents. This allows us to select or ignore features that are of interest.</p>
<p>You can find the description of each feature <a href="http://www.sequenceontology.org/browser/obob.cgi">here</a>.</p>
<p><a id="other-inputs"></a></p>
<h3 id="other-inputs-that-can-be-used-for-gene-set-prs-using-prset">Other inputs that can be used for gene-set PRS using PRSet</h3>
<h4 id="browser-extensible-data-bed">Browser Extensible Data BED</h4>
<p>Browser Extensible Data (BED) file (diÔ¨Äerent to the binary ped file from PLINK), is a file format to define genetic regions. It contains 3 required fields per line (chromosome, start coordinate and end coordinate) together with 9 additional optional field. A special property of BED is that it is a 0-based format, i.e. chromosome starts at 0, as opposed to the usual 1-based format such as the PLINK format. For example, a SNP on chr1:10000 will be represented as:</p>
<table>
<thead>
<tr>
<th style="text-align: center;"></th>
<th style="text-align: center;"></th>
<th style="text-align: center;"></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;"><strong>1</strong></td>
<td style="text-align: center;"><strong>9999</strong></td>
<td style="text-align: center;"><strong>10000</strong></td>
</tr>
</tbody>
</table>
<hr />
<blockquote>
<p>‚ùì How should we represent the coordinate of rs2980300 (chr1:785989) in BED format?
</p>
</blockquote>
<hr />
<h4 id="list-of-snps">List of SNPs</h4>
<p>Finally, PRSet also allow SNP sets, where the user have flexibility to decide what SNPs are included. The list of SNPs can have two different formats:</p>
<ul>
<li>SNP list format, a file containing a single column of SNP ID. Name of the set will be the file name or can be provided using <code>--snp-set File:Name</code></li>
<li>MSigDB format: Each row represent a single SNP set with the first column containing the name of the SNP set.</li>
</ul>
<p><a href="#top"><a href="#table-of-contents">Back to Top</a></a></p>
<p><a id="considerations"></a></p>
<h2 id="considerations-when-analysing-and-interpreting-gene-set-prss">Considerations when analysing and interpreting gene-set PRSs</h2>
<p><a id="clumping"></a></p>
<h3 id="clumping-in-gene-set-prs-analyses">Clumping in gene set PRS analyses</h3>
<p>In standard clumping and P-value thresholding methods, clumping is performed to account for linkage disequilibrium between SNPs. If genome-wide clumping is performed at the gene-set level, we may remove signal as <a href="https://choishingwan.github.io/PRSice/prset_detail/#snp-set-files">shown in this toy example</a>.</p>
<p>To maximize signal within each gene set, clumping is perforemd for each gene set separately.</p>
<p><a id="thresholding"></a></p>
<h3 id="p-value-thresholding-in-gene-set-prs-analyses">P-value thresholding in gene set PRS analyses</h3>
<p>PRSet default option is to no not perform p-value thresholding. It will simply calculate the set based PRS at P-value threshold of 1. </p>
<hr />
<blockquote>
<p>‚ùì Why do you think that the default option of PRSet is P-value threshold of 1?</p>
<p>‚ùì In what cases would you like to apply P-value thresholding? What commands would you use for that?
</p>
</blockquote>
<hr />
<p><a id="p-value-testing"></a></p>
<h3 id="self-contained-vs-competitive-testing">Self-Contained vs Competitive Testing</h3>
<p>An important aspect when calculating gene-set based PRSs is the type of test used for association. Since we are only considering one region of the genome, self-contained and/or competitive tests can be performed. </p>
<p>The null-hypothesis of self-contained and competitive test statistics is diÔ¨Äerent:
  ‚Äì <strong>Self-Contained</strong> - None of the genes within the gene-set are associated with the phenotype
  ‚Äì <strong>Competitive</strong> - Genes within the gene-set are no more associated with the phenotype than genes outside the gene-set
Therefore, a bigger gene-set will have a higher likelihood of having a significant P -value from self-contained test, which is not desirable.</p>
<p><a href="#top"><a href="#table-of-contents">Back to Top</a></a></p>
<p><a id="exercise-4-gene-set-based-prs-analysis"></a></p>
<h2 id="exercise-calculate-gene-set-prs-analysis">Exercise: Calculate gene-set PRS analysis</h2>
<p>We are now ready to perform gene-set association analyses using PRSet.</p>
<p>To perform the PRSet analysis and obtain the set based PRS and competitive P-value, simply provide the GTF file and the GMT file to PRSice and specify the number of permutation for competitive P-value calculation using the --set-perm option.</p>
<div class="highlight" style="background: #f8f8f8"><pre style="line-height: 125%;"><span></span><code>Rscript ./Software/PRSice.R \
    --prsice Software/PRSice_linux  \
    --base Base_Data/GIANT_Height.txt \
    --target Target_Data/TAR \
    --A1 Allele1 \
    --A2 Allele2 \
    --snp MarkerName \
    --pvalue p \
    --stat b \
    --beta \
    --binary-target F \
    --pheno Target_Data/TAR.height \
    --cov Target_Data/TAR.covariate \
    --out Results/Height.set \
    --gtf Reference/Homo_sapiens.GRCh38.86.gtf \
    --wind-5 5kb \
    --wind-3 1kb \
    --msigdb Reference/Sets.gmt \
    --multi-plot 10 \
    --set-perm 1000
</code></pre></div>
<blockquote>
<p><strong>Figure 1.8: An example of the multi-set plot. Sets are sorted based on their self-contained R2 . Base is the genome wide PRS</strong></p>
<p><img alt="Figure 1.8" src="/images/day3/Height.set_MULTISET_BARPLOT_2023-06-30.png" /></p>
</blockquote>
<hr />
<blockquote>
<p>üìå If the --wind-5 and --wind-3 flag is not specified, PRSet will use the exact coordinates of each gene as the boundary. By specifying eg. --wind-5 5kb and --wind-3 1kb then the boundary of each gene will be extended 5 kb towards the 5‚Äô end and 1 kb towards the 3‚Äô end so that regulatory elements of the gene can be included.
</p>
</blockquote>
<hr />
<blockquote>
<p>‚ùì Can you plot the relationship between the gene-set R2 and the number of SNPs in each gene-set? What general trend can be seen?</p>
<p>‚ùì Considering the plot, what gene-sets do you think are most interesting and why?</p>
<p>‚ùì Why is it useful to have polygenic scores measured across gene-sets (or pathways) for individuals? Isn‚Äôt it suÔ¨Écient to just obtain a ranking of gene-sets according to GWAS-signal enrichment?</p>
<p>‚ùì What are the self-contained P-value of the 3 most interesting gene-sets? And what are the competitive P-values? </p>
<p>‚ùì Which P-value would you use to evaluate the importance of gene-set? 
</p>
</blockquote>
<hr />
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/tadesouaiaia/prsWorkshop-website" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../javascripts/details.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
